internal/rpcserver/rpcserver.go

```diff
@@ -5855,19 +5840,27 @@ func (s *Server) Run(ctx context.Context) {
 					s.ntfnMgr.NotifyWork(templateNtfn)
 
 				case <-ctx.Done():
-					s.wg.Done()
+					wg.Done()
 					return
 				}
 			}
 		}(s, ctx)
 	}
 
+	// Run the notification manager and wait for it to terminate.
 	s.ntfnMgr.Run(ctx)
-	err := s.shutdown()
-	if err != nil {
-		log.Error(err)
-		return
+
+	// Close all listeners and wait for all goroutines to terminate.
+	log.Warnf("RPC server shutting down")
+	for _, listener := range s.cfg.Listeners {
+		err := listener.Close()
+		if err != nil {
+			log.Errorf("Problem shutting down rpc: %v", err)
+			return
```

Comment:
Should this be `continue` instead of `return`?  Even if an error occurs here, the full set of listeners should be closed and the `wg.Wait()` below should still be respected

Reply:
Agree, although a `continue` statement isn't needed either, it can just log and move onto the next listener since this is the last part of the loop body.

Reply:
That could potentially cause shutdown to hang.  If a listener fails to close, the waitgroup for that goroutine will never be decremented.

Reply:
That said, I do agree it should ideally still try to close the others.  I'll go ahead and keep track of any of them had an error and only wait when there were none (which is almost always going to be the case of course).

Reply:
Done.